client/src/hooks/useAppNavigation.tsx

```diff
@@ -0,0 +1,142 @@
+import {
+  createContext,
+  useCallback,
+  useContext,
+  useMemo,
+  useState,
+} from 'react';
+import { useNavigate } from 'react-router-dom';
+import { usePersistentState } from './usePersistentState';
+
+interface NavigationItem {
+  type: 'search' | 'repo' | 'full-result';
+  query?: string;
+  repo?: string;
+  path?: string;
+  page?: number;
+  loaded?: boolean;
+}
+
+type ContextType = {
+  navigationHistory: NavigationItem[];
+  navigatedItem?: NavigationItem;
+  navigate: (
+    type: NavigationItem['type'],
+    query?: string,
+    repo?: string,
+    path?: string,
+    page?: number,
+  ) => void;
+  navigateBack: () => void;
+  navigateRepoPath: (repo: string, path?: string) => void;
+  navigateSearch: (query: string, page?: number) => void;
+  navigateFullResult: (repo: string, path: string) => void;
+  query: string;
+};
+
+const AppNavigationContext = createContext<ContextType>({
+  navigationHistory: [],
+  navigate: (type) => {},
+  navigateBack: () => {},
+  navigateRepoPath: (repo, path) => {},
+  navigateSearch: (query, page) => {},
+  navigateFullResult: (repo, path) => {},
+  query: '',
+});
+
+export const AppNavigationProvider = (prop: {
+  value?: string;
+  children: JSX.Element | JSX.Element[];
+}) => {
+  const [navigation, setNavigation] = usePersistentState<NavigationItem[]>(
+    [],
+    'navigation',
+  );
+  const navigateBrowser = useNavigate();
+
+  const navigatedItem = useMemo(
+    () => (navigation.length ? navigation[navigation.length - 1] : undefined),
+    [navigation],
+  );
+
+  const buildQuery = (navigationItem: NavigationItem) => {
+    const { query, path, repo, type } = navigationItem;
+    switch (type) {
+      case 'repo':
+      case 'full-result':
+        return `open:true repo:${repo} ${path ? `path:${path}` : ''}`;
+      case 'search':
+        return query;
```

Comment:
I think we should encodeURIComponent here to fix issues with having special symbols in the url